### DISCUSSION TITLE
Functions only show functions for the `public` schema

### DISCUSSION THREAD

**Discussion by bombillazo** [supabase/supabase#16955] [Contribute to Supabase]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

In the Supabase dashboard, when one creates a custom function for a schema other than `public`, they are created but not shown in the dashboard. We confirm that they exist by running:

```sql
SELECT
    routine_name
FROM 
    information_schema.routines
WHERE 
    routine_type = 'FUNCTION'
AND
    routine_schema = 'auth';
```

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. In the Supabase dashboard, go to 'Database -> Functions'
2. Click on 'Create new function'
3. In the schema, select a schema beside `public` (for example, the `auth` schema)
4. Fill in the rest of the 'Add a new function' form
5. In the functions dashboard, the function will not appear.
6. Run the SQL query above to confirm the function exists.

## Expected behavior

I expect to see functions created in the non-`public` schema

## Screenshots

Only the public schema functions are shown in the dashboard.

<img width="1502" alt="image" src="https://user-images.githubusercontent.com/44002321/228097703-1140472e-6c47-4282-80bf-3bce72bd4f3a.png">

## System information

- OS: All

**Comment by nrayburn-tech** [comment]
The functions that are displayed in the UI are filtered by a list of schemas.  If this filter is removed, then all of the default available functions are displayed in the UI (something like 200+ from my test default project).
https://github.com/supabase/supabase/blob/4ed3e2620c3dec574d3c0140bd9c22b6d30fdee0/studio/stores/pgmeta/MetaStore.ts#L172-L189

To remove the filter, updates need to be made to the below two places.
https://github.com/supabase/supabase/blob/4ed3e2620c3dec574d3c0140bd9c22b6d30fdee0/studio/components/interfaces/Database/Functions/FunctionsList/FunctionsList.tsx#L28-L30
https://github.com/supabase/supabase/blob/4ed3e2620c3dec574d3c0140bd9c22b6d30fdee0/studio/components/interfaces/Database/Functions/FunctionsList/FunctionList.tsx#L27

There are a few options on how to fix this that I came up with below, but somebody from the Supabase team should weigh in on how it should be handled.
- Show all the functions, including the default functions.  Can be done by just simply removing the filtering of excluded schemas.
- Show only user created functions.  Would need some way to identify a function as user created, which would likely require some API change to the `/api/pg-meta/default/functions` endpoint so that it returns whether a function is user created or not.
- Hide the excluded schemas from the dropdown when creating functions.  This way the user can only add functions to schemas that are shown in the functions list.

**Comment by bombillazo** [comment]
Hey @nrayburn-tech, thanks for the deep dive! A good solution would be to have a UI control that can dynamically update the filters so that not one is hardcoded and it gives users the options. The possible filters could be to show user-created schemas, default, or all.

**Comment by KTayy** [comment]
``
select n.nspname as schema_name,
       p.proname as specific_name,
       case p.prokind
            when 'f' then 'FUNCTION'
            when 'p' then 'PROCEDURE'
            when 'a' then 'AGGREGATE'
            when 'w' then 'WINDOW'
            end as kind,
       l.lanname as language,
       case when l.lanname = 'internal' then p.prosrc
            else pg_get_functiondef(p.oid)
            end as definition,
       pg_get_function_arguments(p.oid) as arguments,
       t.typname as return_type
from pg_proc p
left join pg_namespace n on p.pronamespace = n.oid
left join pg_language l on p.prolang = l.oid
left join pg_type t on t.oid = p.prorettype
where 
n.nspname not in ('pg_catalog', 'information_schema') and
p.proname = 'Name_of_your_function'
order by schema_name,
         specific_name;
`

Hey, not sure if this is still an open question. I ran into a similar issue and this helped me. you can just copy the definition into an sql editor to get a readable version. If you do not know the name of your function just remove the p.proname where clause.

**Comment by robpetty** [comment]
Is there any update on this?  I actually totally forgot about this problem so trying to be lazy I was standing up a new project, creating a multi-tenant setup, adding functions and triggers and realized I can't see them.

@KTayy I had to make a small tweak to see my functions, your OG code there says no rows found for me so had to make some tweaks:

> SELECT 
    n.nspname AS schema_name,
    p.proname AS specific_name,
    CASE p.prokind
        WHEN 'f' THEN 'FUNCTION'
        WHEN 'p' THEN 'PROCEDURE'
        WHEN 'a' THEN 'AGGREGATE'
        WHEN 'w' THEN 'WINDOW'
    END AS kind,
    l.lanname AS language,
    CASE 
        WHEN l.lanname = 'internal' THEN p.prosrc
        ELSE pg_get_functiondef(p.oid)
    END AS definition,
    pg_get_function_arguments(p.oid) AS arguments,
    t.typname AS return_type
FROM 
    pg_proc p
LEFT JOIN 
    pg_namespace n ON p.pronamespace = n.oid
LEFT JOIN 
    pg_language l ON p.prolang = l.oid
LEFT JOIN 
    pg_type t ON t.oid = p.prorettype
WHERE 
    n.nspname = 'your_schema_name'  -- Filter for the my_cool_schema
ORDER BY 
    schema_name,
    specific_name;
    
Also, to anyone just wanting to see it all, just comment out the WHERE clause and you will get all the schemas and all the functions:

> -- WHERE 
> --    n.nspname = 'your_schema_name'  -- Filter for the my_cool_schema

**Reply by GaryAustin1** [reply]
The UI shows triggers and functions by setting the schema in both tabs.

**Reply by robpetty** [reply]
Hey @GaryAustin1 yep.  I stand corrected but it was odd.  Best I can tell it took multiple minutes before it was reflected in the UI.  I could query it per my SQL above and see it (the function).  But seriously, was maybe 5 minutes after I gave up messing around and poof there it was (I had set the schema to the correct one, its how I saw initially it was placing them in the wrong one (public) when I tried manually (without schema) and via the AI...even though in both cases during creation I had the schema I wanted selected).
